[1단계 - 자동차 경주 구현] 후추(주찬민) 미션 제출합니다.

Combi153 opened the pull request:
안녕하세요! 반갑습니다. 우아한 테크코스 5기 후추입니다.

코드를 작성하면서 고민했던 사항과 어려움을 남겨봅니다. 리뷰 시 고려해주신다면 정말 감사하겠습니다!

1. 시간 부족으로 인한 하드코딩
 - 시간이 부족하여 하드코딩된 부분이 많습니다. 2단계 미션을 진행하면서 하드코딩 등을 수정해보도록 하겠습니다.
 
2. 좋은 구조란 무엇인가?
- MVC 패턴에 대한 얕은 지식으로 해당 구조를 사용하였습니다. 그러나 같은 MVC 패턴임에도 사람마다 그를 구현하는 방법이 다른 것 같습니다. 그러다보니 단순히 어떤 패턴을 사용하는 것이 좋은 구조의 조건이라고 말하기 어려운 것 같습니다. 좋은 구조를 갖기 위해 어떤 것들을 고려해야 하는지 궁금합니다. 

3. getter 를 사용하지 않으려 노력했지만, 저희가 설계한 구조에서는 getter 사용이 불가피하다고 느꼈습니다. 어떤 방식으로 리팩토링하면 좋을지 힌트를 주신다면 감사하겠습니다.

4. 자바를 자바답게 사용하지 않은 곳이 있다면 모두 알려주신다면, 열심히 공부해보겠습니다.

감사합니다. 열심히 성장하겠습니다!

the9kim committed:
docs(README) 기능 목록을 작성한다

the9kim committed:
feat: 자동차 이름을 입력받는 기능을 구현한다

the9kim committed:
feat: View에 게임 시도 횟수를 입력받는 기능을 구현한다

the9kim committed:
feat: CarFactory 객체에 Car 객체와 CarRepository 객체를 생성하는 기능을 구현한다

the9kim committed:
feat: 랜덤 숫자를 생성하는 기능을 구현한다

the9kim committed:
feat: 자동차를 이동하는 기능을 구현한다

the9kim committed:
feat: 자동차의 이름을 검증하는 기능을 구현한다

the9kim committed:
feat: 자동차 이름의 개수 및 중복 여부를 검증하는 기능을 구현한다.

the9kim committed:
feat: 실행 결과 및 최종 우승자 출력 기능을 구현한다

the9kim committed:
feat: 입력값 공백 여부를 검증하는 기능을 구현한다

the9kim committed:
feat: 게임 우승자를 찾는 기능을 구현한다

the9kim committed:
feat: Application 객체에 게임을 실행하는 기능을 구현한다

the9kim committed:
feat: 전체 게임에 관한 Controller 를 구현한다

the9kim committed:
fix: 사용하지 않는 객체를 삭제한다

the9kim committed:
fix: 필드의 접근제어자를 변경한다

the9kim committed:
test: Car 객체의 updatePosition 기능을 테스트한다

the9kim committed:
test: Car 객체의 생성 및 검증 기능을 테스트한다

the9kim committed:
test: CarFactory의 CarRepository를 생성하는 기능을 테스트한다

the9kim committed:
refactor: CarRepository의 movePosition 메소드를 리팩토링한다

stream 연산으로 수정한다.

the9kim committed:
test: CarRepository의 movePosition 메소드를 테스트 한다

the9kim committed:
test: CarRepository 의 생성 및 검증 기능을 테스트한다

the9kim committed:
test: CarRepository 의 findWinner 기능을 테스트 한다

the9kim committed:
refactor: CarRepository 객체의 movePosition 메소드를 리팩토링한다

변경 전:
난수를 외부에서 생성하여 파라미터로 받았다.

변경 후:
메소드 내에서 난수를 생성하도록 변경한다.

변경 이유:
자동차 객체별로 서로 다른 난수를 적용하기 위함이다.

the9kim committed:
refactor: GameController 의 moveCar 메소드를 리팩토링한다

변경 전:
moveCar 메소드 내에 자동차를 움직이는 비즈니스 로직이 존재한다.

변경 후:
해당 비즈니스 로직을 GameSystem 객체 내에 옮겨 구현한다.

the9kim committed:
chore: InputView 메소드들의 세부 내용을 정리한다

the9kim committed:
refactor: printResult 메소드의 파라미터가 변경됨에 따라 함수 내용을 리팩토링한다

the9kim committed:
refactor: 게임 결과에 대한 데이터를 Result 객체 내에 담도록 리팩토링한다

the9kim committed:
refactor: carRepository 객체의 movePosition 메소드를 리팩토링한다

the9kim committed:
refactor: Car 객체에 equals, hashCode 메소드를 구현한다

the9kim committed:
test: Result 의 우승자를 찾는 기능을 테스트한다

the9kim committed:
test: GameSystem 의 자동차 경주를 실행하는 기능을 테스트한다

the9kim committed:
fix: 스트림의 연산 내용을 추가 및 수정한다

the9kim committed:
fix: printResult 메소드를 분리한다

Combi153 committed:
refactor: OutputView 객체 내에 하드코딩을 없애고 메소드를 리팩토링한다

변경 전:
각 메소드 내에 출력 메시지들이 존재한다.

변경 후:
Message 객체를 만들어 출력 메시지를 관리한다
메소드를 분리한다.

변경 이유:
출력 메시지를 따로 관리하는 것이 효율적이기 때문이다.

Combi153 committed:
refactor: InputView 클래스를 리팩토링한다

변경 내용:
readCars, line, validateLine 등 메소드와 변수의 이름을 변경한다.
readGameRound 메소드에서 입력값이 숫자인지 검증하는 기능을 추가한다.

Combi153 committed:
refactor: view 계층에서 에러 메시지를 관리하는 ErrorMessage 객체를 새로 만들어 InputView에서 사용한다

Combi153 committed:
refactor: GameController 객체의 메소드를 다시 정리 및 분리한다

Combi153 committed:
fix: CarFactory 객체의 createCar 메소드를 삭제하고 Car 생성자를 사용한다

Combi153 committed:
fix: 실행결과 안내 메시지를 printResult 메소드에서 출력할 수 있도록 한다

Combi153 committed:
fix: 메서드 이름을 수정하고, 매직넘버 등을 상수로 선언한다

Combi153 committed:
chore: 패키지명과 객체 이름을 수정한다

변경 전:
carrepository, CarRepository

변경 후:
cars, Cars

Combi153 committed:
fix: findWinners 테스트 메소드의 내용을 수정한다

Combi153 committed:
fix: validate_test 메소드의 내용을 수정한다

Hyeon9mak commented on docs/README.md:
```diff
@@ -0,0 +1,28 @@
+                    - 자동차 간 이름이 중복일 경우
+            - 시도 횟수
+                - 공백
+                - 정수x (1이상, Integer 범위)
```
`- [ ]`, `- [x]` 를 사용하면 체크박스를 생성해서 어떤 작업을 얼마나 진행했는지 기록할 수 있어요! 👍

Combi153 replied:
오 알려주셔서 감사합니다 적극 활용해볼게요!

Hyeon9mak commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,45 @@
+    private Cars makeCars(List<String> carNames) {
+        CarFactory carFactory = new CarFactory();
+        return carFactory.createCarRepository(carNames);
+    }
```
호출되는 순서에 따라 메서드를 정리해주면 가독성이 더 좋아질거 같아요 😄

Combi153 replied:
앗 꼼꼼하지 못했네요 앞으로는 더 신경써보겠습니다!

Hyeon9mak commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,64 @@
+
+    public static final String ENGLISH_FORMAT = "[a-zA-Z]+";
+    public static final int MAXIMUM_ROUND = 5;
+    public static final int MINIMUM_ROUND = 1;
```
`Car` 내부에서만 사용되는 상수들은 private 접근제한자로 가려줄 수 있겠죠?

Combi153 replied:
이 역시 마찬가지네요ㅠㅠ 단축키를 활용해 상수를 만들고 미처 확인을 안 했습니다 얼른 고칠게요~!

Hyeon9mak commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private void validate(String name) {
+        if (!isValidName(name)) {
```
부정형 표현을 위해 if 키워드 내부에서 사용되는 `!`는 잘 눈에 띄지 않는데요, `isValidName` 표현을 `isNotValidName` 등으로 바꾸어주면 문제를 해결해줄 수 있어요!

Combi153 replied:
언젠가 함수 네이밍에 대해 알아보고 고민할 때, boolean 값을 반환하는 함수에 `부정형` 표현을 쓰지 않는 것이 좋다는 생각을 하였습니다. 왜냐하면 함수의 반환값을 예상할 때, 부정형으로 인해 뒤집어서 생각해야 되니까 헷갈릴 수 있겠더라구요. 그래서 `!`을 지향하려 했습니다.

현구막님 말씀을 듣고 보니, 헷갈리기란 둘 다 마찬가지일 거라는 생각이 드네요! 오히려 눈에 띄도록 `not` 등을 붙이는 것이 좋아보입니다.

한편, 헷갈리지 않으면서도 `not`을 붙이지 않는 방법은 없을까요? `isInvalidName` 정도는 괜찮은 이름일까요?

Hyeon9mak replied:
그런 고민을 거치셨군요! 💯 
`!` 를 사용하는 것과 `not`을 사용하는 것 모두 가독성 가시성이라는 트레이드오프가 있으니 후추가 생각하시기에 더 중요하게 느껴지는 걸 선택하시면 될 것 같습니다.

Combi153 replied:
저는 `valid`의 반대 표현인 `invalid`가 있으니 `isInvalidName` 으로 정하는 것이 좋을 것 같습니다.

트레이드 오프 상황을 마주했을 때 근거 있는 선택을 내리는 연습을 많이 해보겠습니다!!

Hyeon9mak commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,64 @@
+
+    public void updatePosition() {
+        this.position++;
+    }
```
조금 더 자동차 경주 게임에 어울리는 행위를 네이밍으로 정해주면 어떨까요? 😄 
또 그렇게 했을 때 생기는 장점은 무엇일까요? (유지보수 관점)

Combi153 replied:
리뷰를 받은 후 `객체지향`에 대해 다시 돌아보는 중입니다. 감사드려요!😀

- 이번에 진행한 미션 내용이 자동차 경주 게임이고, 이 기능은 `자동차` 객체가 가지므로 updatePosition 보다는 `move` 정도가 어떨까 해요. 앞으로는 객체가 주체성을 갖도록 네이밍에 신경을 쓰겠습니다!

- `move` 등으로 자동차 객체와 어울리는 네이밍을 했을 때 다음과 같은 장점이 있을 것 같아요

1. 보다 추상화된 이름이에요. 구현 기능이 변경되더라도 이름 덕분에 새로운 함수를 만들지 않아도 될 것 같아요. 가령, move 메소드 내에 "여러 번" 움직이거나, "앞으로 혹은 뒤로" 움직이는 등의 기능이 추가되어도 이상하지 않아요
2. 가독성이 좋아졌어요. 다른 개발자들이 이해하기 편한 이름이에요. Car 객체를 외부에서 사용하거나, Car 객체를 직접 수정할 때 모두 도움이 될 것 같아요

- 놓친 장점이 있다면 말해주세요. 또 고민해볼게요!

Hyeon9mak replied:
1번 답변을 기대하고 드린 질문이었습니다 😄 👍 💯

Hyeon9mak commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, position);
+    }
```
동등성 부여 👍 
실제 사용처는 `이름 중복 검증`인거 같은데요, position 까지 같아야 할 필요가 있을까요?

Combi153 replied:
position 까지 같을 필요는 없을 것 같아요!

자동차 객체 간 이름의 중복이 허용된다면 필요할 수도 있을 것 같다는 생각이 순간 머릿속을 스쳐지나갔지만,,, 그때 역시 position으로 구분하는 것이 완전한 구분이라고 볼 수는 없겠네요!

이런 점에서도 세심하게 코드를 짜는 연습을 더 해볼게요 감사합니다😁

Hyeon9mak replied:
만약 이름중복을 허용한다면 name + position 으로 동등성을 부여하는게 아니라, 완전히 equals & hashCode 오버라이드를 제거하고 인스턴스 고유의 해시값으로 비교되도록 바꿔야할 것 같아요. 현재는 이름을 통한 비교가 의미가 있으므로 `name` 으로만 동등성을 부여해주면 될 것 같습니다.

Combi153 replied:
넵 알겠습니다! 인스턴스 해시값으로 비교하는 것은 다음에 실험해보고 싶네요

Hyeon9mak commented on src/main/java/racingcar/domain/carfactory/CarFactory.java:
```diff
@@ -0,0 +1,20 @@
+        return carNames.stream()
+                .map(Car::new)
+                .collect(Collectors.toList());
+    }
```
메서드 명칭은 `createCareRepository` 인데 실제 반환타입은 `Cars`인거 같아요! 네이밍을 조금 더 직관적으로 바꾸어주면 어떨까요?

Combi153 replied:
앗! Cars 객체의 처음 이름이 CarRepository 였어요. 객체 이름을 Cars 로 바꾸면서 놓쳤네요ㅠㅠ 이름을 바꾸는 것도 신중해야할 것 같습니다!
말씀하신 메소드의 이름이 `createCars` 라면 `직관적`인 이름에 부합할까요?

Hyeon9mak replied:
네 그정도면 좋은 것 같아요 😄

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+
+    private static final int MOVING_CONDITION = 4;
+    private final List<Car> repository;
+
```
일급 컬렉션 활용 👍

Hyeon9mak replied:
`MOVING_CONDITION` 는 현재 사용되지 않는 것 같아요.

Combi153 replied:
얼른 지우겠습니다!

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+        if (hasDuplication(repository)) {
+            throw new IllegalArgumentException("[ERROR] 자동차 이름은 중복될 수 없습니다.");
+        }
+    }
```
중복된 자동차가 경주에 포함되어선 안되는군요! 혹시 `Set` 자료구조를 사용해볼 수도 있지 않을까요? `List`와 `Set` 은 어떤 차이가 있을까요? `List`가 더 나을까요? 후추의 생각이 궁금합니다 😄

Combi153 replied:
- `List`는 중복을 허용하지만, `Set`은 중복을 허용하지 않는다는 특징이 있습니다.

- 자동차 간 중복이 있다면 오류가 발생하도록 검증 기능을 구현해두었기 때문에 처음에는 `Set` 자료구조의 필요성을 느끼지 못했습니다.

- 하지만 그러한 정보는 `Cars 클래스의 코드를 열어봐야만 알 수 있는` 정보입니다. 만약 `Set` 구조를 사용했다면 다른 객체에서 Cars의 필드를 사용할 때 중복이 허용되지 않은 것을 유추할 수 있겠어요!

- 생성 및 검증 단계에서 중복을 막고, 이후 자료의 특성으로도 중복을 막는 것이 객체의 주기 및 유지보수 관점에서도 긍정적인 방향일 것 같습니다.

- 한편 자동차 간 중복이 있다면 오류가 발생하도록 하는 것은 여전히 구현하고픈 기능입니다. 따라서 생성자에서 입력을 받을 때는 `List`로 받아 검증을 받도록 하고, 이후 `Set`자료구조로 초기화하도록 리팩토링 하는 것이 좋을 것 같습니다.

- `List`를 사용할 때와 달리, `Set`을 사용하니 입력 데이터에 대한 순서가 보장되지 않는 특징을 발견했습니다. `LinkedHashSet`을 사용하면 되나 성능면에서 뒤쳐지지는 않을까 우려됩니다. 혹시 어떻게 생각하시는지 궁금합니다!

깊게 생각해볼 수 있는 또 하나의 키워드였네요 감사드립니다😁😁😁

Hyeon9mak replied:
> 생성자에서 입력을 받을 때는 List로 받아 검증을 받도록 하고, 이후 Set자료구조로 초기화하도록 리팩토링 하는 것이 좋을 것 같습니다. 

여기까지도 생각하신 점 너무 멋집니다. 기대했던 모든 답변이 다 나왔어요!!!! 💯 💯 💯 💯 
`LinkedHashSet` 까지는 저도 생각하지 못했네요 👍 👍 👍 
코멘트의 의도는 '비즈니스 로직에 어울리는 자료구조가 있지는 않을지 항상 고민하기' 였습니다.

- 말씀주신대로 `List`는 중복을 허용해서 자체적인 검증이 필요하고, `Set`은 순서를 보장해주지 않아 출력순서를 보장할 수 없습니다.
- `LinkedHashSet`이 가장 적합해보이는데, 성능이 우려됩니다. 
- 하지만 현재 자동차 경주 미션에 참가자가 999,999,999명 동시에 참여할 일이 있을까요?
- 보편적인 `List`를 놔두고 `LinkedHashSet`를 써야할만큼 중요한 로직일까요?

모든 기술은 트레이드오프가 있고 걔중에 애플리케이션 특성에 가장 부합하는(가장 단점이 적은)걸 선택하면 되는 것 같습니다.
애플리케이션 특성에 맞는 건 무엇일지, 팀원들이 유지보수할 때 어떻게 느낄지를 고려해보시고 후추만의 결정을 내려주세요. 그리고 저를 설득해주세요 😄

Combi153 replied:
## List 사용에 대한 생각
- 출력순서가 보장됩니다.
- 중복을 허용합니다.
- 중복 여부에 관한 정보를 전달하지는 못합니다.
- 보편적인 자료형으로 다른 개발자들이 사용할 때 어려움이 없습니다.

## Set 사용에 대한 생각
- 출력순서가 보장되지 않습니다.
- 중복을 허용하지 않습니다.
- 중복 여부에 관한 정보를 간접적으로 전달합니다.
- 출력 순서를 보장하기 위해 `LinkedHastSet` 등을 활용해야 합니다. 이는 성능 면에서 떨어집니다.
- 다른 개발자가 `LinkedHashSet`이 아닌 다른 Set 자료형으로 새로운 게임을 개발할 시 의도대로 동작하지 않을 수 있습니다.

## 결론
- 두 자료형에 대해 `비즈니스 로직에 더 어울리는 자료구조 무엇일지` 고민을 해본다면, `자동차 간의 순서가 반드시 보장되어야 할 것` 같습니다. 
- 이때 Set 자료형을 쓴다면 유지 보수 시 다른 개발자들에게 혼란을 초래할 수 있을 것 같습니다. 따라서 `List`를 사용하는 것이 맞다는 생각이 듭니다! List로 수정해보겠습니다.
- 다만 중복 검증 이후 새로운 데이터가 추가되면 의도하지 않은 중복 데이터가 생성될 우려가 있습니다. 따라서 Collections.unmodifiableList()방법을 적극 활용하겠습니다.

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+                        Car::getPosition,
+                        (oldValue, newValue) -> oldValue,
+                        LinkedHashMap::new));
+    }
```
`Cars` 는 경주에 참가한 자동차들 묶음(컬렉션)으로 자동차들이 일괄적으로 수행해야할 비즈니스로직을 담당하는 객체 같은데요, 현재 `movePosition()` 메서드는 경주 수행 결과가 어떻게 출력(이름, 위치)될지를 인지(`Map<String, Integer>`)하고 있는 것 같아요. 이럴 경우 어떤 문제가 생길까요?

Combi153 replied:
와 이런 생각은 못해봤습니다! 리뷰 최고입니다 현구막!!!😆😆😆

- 출력에 대한 요구사항이 변경된다면 문제가 생길 것 같습니다. 
출력 방식을 변경할 때는 View의 코드를 바꾸는 것이 마땅한데, 비즈니스 로직을 담당하는 코드까지 바꾸는 일이 생길 것 같아요! Domain 에서 지나치게 많은 정보를 알고 있는 상황 같습니다.
- 데이터를 있는 그대로 전달하는 것이 좋은 방법이 될 것 같습니다!

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+        if (randomNumber >= 4) {
+            car.updatePosition();
+        }
+        return car;
```
매직넘버! `MOVING_CONDITION` 로 대체 가능하겠네요 😄

Combi153 replied:
맞습니다!! 꼼꼼한 코드 작성에 유의할게요!

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+
+    public List<Car> getRepository() {
+        return repository;
+    }
```
사용되지 않는 메서드는 동료 개발자들에게 혼란을 줄 수 있어요 🙏

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -0,0 +1,73 @@
+        int maxPosition = repository.stream()
+                .map(Car::getPosition)
+                .max(Comparator.comparing(x -> x))
+                .orElseThrow(NoClassDefFoundError::new);
```
- 현재 메서드는 테스트에서만 사용되고 있는 것 같아요 😱 
- `NoClassDefFoundError` 는 컴파일 타임에 존재했던 클래스에 런타임에 존재하지 않을 때 발생하는 에러입니다. 개발자들이 잘 아는 예외를 사용하는 건 가독성에 도움이 되지만 잘못 사용하면 더 큰 역효과를 만들 수 있어요!

Combi153 replied:
이 기능을 Result 객체로 옮기면서 지우지 않았던 것 같습니다. `NoClassDefFoundError` 는 저도 처음 보는 에러네요🤔 언제 왜 이 코드를 적었는지 모르겠습니다...

꼼꼼한 코드 작성! 정신차리고 코딩하기! 명심하겠씁니다!!!

Hyeon9mak commented on src/main/java/racingcar/domain/numbergenerator/NumberGenerator.java:
```diff
@@ -0,0 +1,11 @@
+        Random random = new Random();
+        return random.nextInt(9);
+    }
+}
```
- `Random().nextInt(9)` 는 0~8 사이의 난수가 발생됩니다.
- 명칭만 보아서는 0~9 사이의 랜덤한 숫자를 발급한다는 걸 알기가 어려울 것 같아요! 메서드 네이밍을 조금 다듬어주면 어떨까요? 😉

Combi153 replied:
- 메소드의 사용법을 보다 자세히 알고 사용하는 습관을 들여야 할 것 같아요! 알려주셔서 감사드립니다!

- 혹은 테스트코드에서 0부터 9가 나오는지 테스트하였다면, 이런 실수를 방지할 수 있었을 것 같네요. Random한 값을 불러오는 기능인지라 테스트 코드 작성에 어려움이 있어 못했습니다. Random 값에 대한 테스트 코드를 짜는 방법도 존재하나요? 존재한다면 도전해보고 싶습니다!

- 메서드 네이밍도 고려해보겠습니다!

Hyeon9mak replied:
랜덤한 값을 테스트 하는 방법은 여러가지가 있을 것 같은데, 그전에 `Random` 라이브러리를 꼭 테스트 해야하는가 고민 해봐야 할 것 같아요. 후추는 `List`, `Set`, `Map`, `String`, `Integer` 와 같은 자료구조, 라이브러리들을 테스트하고 사용하시나요? 이것들을 믿고 사용하시지 않나요?

자신이 불안을 느끼는 요소에 대해 테스트를 작성하는게 좋지만 테스트 작성도 결국 리소스고, 테스트 하나하나도 유지보수 포인트가 됩니다. 저명한 라이브러리나 프레임워크 같은 경우 사용설명서(공식문서)를 제공해주니 이를 참고하는 습관을 기르면 많은 시간을 아끼실 수 있을 것 같아요. https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#nextInt-int-

Combi153 replied:
> 테스트 작성도 결국 리소스고, 테스트 하나하나도 유지보수 포인트가 됩니다.

이 문장이 와닿네요. 어쩌면 `테스트 코드 그 자체를 목적`으로 두고 코드를 작성하려고 했던 것 같습니다.

앞으로는 기술을 맹목적으로 추구하기보다는 `기술의 배경과 이유`에 대해 고민하고 활용하는 것이 맞을 것 같습니다.

소중한 교훈을 알려주셔서 감사합니다😄

Hyeon9mak commented on src/main/java/racingcar/domain/result/Result.java:
```diff
@@ -0,0 +1,38 @@
+                .stream()
+                .max(Comparator.comparing(x -> x))
+                .orElseThrow(IllegalArgumentException::new);
+    }
```
`Result` 는 결과 출력을 위해 사용되는 객체인 것 같은데요,
내부에서 우승자를 가리기 위한 비즈니스 로직을 포함하고 있는 것 같아요.

비즈니스 로직 수행과 출력을 온전히 분리할 수 있을까요?

Combi153 replied:
분리해보겠습니다!

Hyeon9mak commented on src/main/java/racingcar/domain/system/GameSystem.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final int gameRound;
+    private final NumberGenerator numberGenerator;
+
```
`NumberGenerator`가 `Cars` 내부에서 생성되어 사용되는 덕분에 `GameSystem`에서는 사용되지 않는 것 같아요 😱

Combi153 replied:
꼼꼼하게 보기! 다시 한 번 가슴 깊이 새깁니다!!!!!!!!

Hyeon9mak commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,50 @@
+        return Integer.parseInt(gameRoundContent);
+    }
+
+    private void validateCarNames(String content) {
+        if (content.isBlank()) {
+            throw new IllegalArgumentException(ErrorMessage.BLANK_INPUT_ERROR_FOR_CAR_NAMES.getMessage());
+        }
+    }
+
+    private void validateGameRound(String content) {
+        if (!content.matches(FORMAT_OF_NUMBER)) {
+            throw new IllegalArgumentException(ErrorMessage.NOT_NUMBER_ERROR_FOR_GAME_ROUND.getMessage());
+        }
+    }
```
오! 공백인지, 숫자인지에 대한 간단한 검증연산은 `View`의 책임으로도 볼 수 있죠 💯 👍

Hyeon9mak commented on src/test/java/racingcar/domain/carfactory/CarFactoryTest.java:
```diff
@@ -0,0 +1,36 @@
+                Arguments.arguments(List.of("poy", "joy", "kane"))
+        );
+    }
+}
```
![image](https://user-images.githubusercontent.com/37354145/218042990-1a6904bc-0f70-4c43-9e21-78ef1801d8c5.png)

https://seongwon.dev/Git/20220303-%ED%8C%8C%EC%9D%BC%EC%9D%98_%EB%A7%88%EC%A7%80%EB%A7%89_%EA%B0%9C%ED%96%89/

참고해주세요 😄

Combi153 replied:
설정했어요😎

Hyeon9mak commented on src/test/java/racingcar/domain/carfactory/CarFactoryTest.java:
```diff
@@ -0,0 +1,36 @@
+
+    @ParameterizedTest
+    @MethodSource("getCarNames")
+    void createCarRepository(List<String> carNames) {
```
무엇에 대한 테스트인지 명시해주면 동료 개발자들이 테스트 의도를 파악하기 더 좋을 것 같아요 👍

Combi153 replied:
DisplayName을 추가하겠습니다!

Hyeon9mak reviewed (changes requested):
안녕하세요 후추! 자동차 경주 미션을 함께 하게된 현구막입니다~
미션 리뷰가 늦었어요 🙏 

MVC 패턴에 대한 고민이 많으시군요! 그래도 미션을 잘 진행해주신 것 같아서 너무 멋집니다 👍 
차근차근 진행해보면 좋을 것 같아서 디테일하게 코멘트를 남겼어요!
미션 진행이 더뎌도 좋으니 천천히 확인해주세요.

> 좋은 구조란 무엇인가? MVC 패턴

너무 좋은 고민!! 맹목적으로 "MVC 패턴 짱이래. 이렇게 짜는게 객체지향이야. 이게 아름다운 코드지." 라는건 아무런 의미가 없다고 생각해요. 정말 왜 좋은지? 왜 객체지향이어야만 하는지? 건강한 고민이 뒤따른다면 복잡해보이는 패턴도 '아 이래서~' 하고 이해가 되는거 같아요.

MVC 패턴을 포함한 모든 디자인 패턴, 아키텍처는 결국 유지보수를 위해서 존재한다고 생각해요. 모든 서비스는 유지보수해야 할 포인트가 다른 것 같아요. 어떤 서비스는 데이터는 잘 변화하지 않는데 브라우저로 출력했다가, 모바일로 출력했다가 뒤바뀌는 경우가 잦고, 또 어떤 서비스는 출력은 항상 모바일 앱으로만 하는데 내부 데이터와 구조가 시시각각 변화해요. 이들은 각각 출력기능을 유지보수하기 쉬운 패턴과 구조를 바꾸기 쉬운 패턴을 적용하는게 좋겠죠?

자동차 경주 게임의 경우 자동차 경주 진행 방식과 화면 출력 방식 중 어떤게 자주 변경될까요?
아무래도 출력방식에 대한 변화가 잦기 때문에 MVC 패턴이 도입된 것 같아요.
그렇다면 출력에 대한 유지보수가 쉽도록 구조를 만들어야 자동차 경주 게임에 잘 어울리는 MVC 패턴을 도입했다고 말할 수 있을 것 같아요. (출력에 대한 유지보수가 쉽다는 건, View 의 변화에 Controller 와 Model 이 영향을 안받는다는 뜻이겠죠? 😉 )

현재 애플리케이션의 동작 특성과 유지보수 할 때 어떤 부분에 가장 주의를 기울여야하는지를 고민하면 자연스럽게 좋은 구조에 대해 떠올리실 수 있을 것 같아요.

> getter 사용

getter 를 사용하면 안될 이유가 있나요? 어떤 점에서 getter 사용을 지양해야한다고 느끼셨나요? 😄 

> 자바답게

저도 제 자신이 '자바답게' 짜는진 모르겠지만... 열심히 코멘트 남겨보았습니다! 코멘트 내용 확인해주세요 😄 👍 

진행하시면서 궁금하신 점 부담없이 답글이나 DM 남겨주세요 🙇‍♂️

Combi153 committed:
docs: 고민과 리팩토링에 대한 내용을 기록한다

Combi153 committed:
chore: GameController 클래스의 메소드 순서를 변경한다

Combi153 committed:
chore: Car 객체의 상수의 접근제어자를 변경한다

Combi153 committed:
chore: Car 객체의 메소드 이름을 변경한다

Combi153 committed:
fix: Car 객체의 equals, hashCode 메소드의 내용을 변경한다

변경 내용:
동등성 부여 시, name을 기준으로 하도록 변경한다.

Combi153 committed:
chore: CarFactory 객체의 메소드 이름을 변경한다

Combi153 committed:
refactor: Cars 클래스를 리팩토링한다

변경 내용 :
- 매직넘버를 상수로 만들어 사용한다.
- 필드의 이름을 cars로 변경한다.
- movePosition 메소드의 이름을 moveBy로 변경한다.
- moveBy 메소드가 NumberGenerator 객체를 인자로 받아 사용하도록 한다.
- getRepository 메소드를 삭제한다.
- findWinner 메소드를 삭제한다.

Combi153 committed:
feat: Car 클래스의 toString 메소드를 추가한다

Combi153 committed:
chore: CarFactory 클래스의 makeCars 메소드의 세부내용을 변경한다

Combi153 committed:
chore: CarFactoryTest 의 메소드 이름을 변경한다

Combi153 committed:
refactor: Cars 객체를 리팩토링한다

변경 내용 :
- 클래스 변수 cars를 Set 자료구조로 변경한다.
- repository 변수의 이름을 cars로 변경한다.
- moveBy 메소드의 반환값을 없앤다.
- getter 메소드를 추가한다.

Combi153 committed:
refactor: CarTest 테스트코드를 리팩토링한다

변경 내용 :
- 메소드의 이름을 수정한다.
- DisplayName 내용을 수정한다.
- Car 객체의 move 기능을 테스트 하는 내용을 수정한다.
power에 따라 움직였는지 여부를 확인할 수 있도록 한다.

Combi153 committed:
refactor: CarsTest 테스트코드를 리팩토링한다

변경 내용 :
- 메소드의 이름을 수정한다.
- DisplayName 내용을 수정한다.
- findWinner 메소드의 테스트 코드를 삭제한다.
- moveBy 메소드에 대해 움직이는 것과 멈춰있는 것 두 가지의 테스트 코드를 추가 및 수정한다.

Combi153 committed:
refactor: GameSystem 클래스를 리팩토링한다

변경 내용 :
- 매직넘버를 상수로 선언하여 사용한다.
- 사용하지 않는 필드를 삭제한다.
- 생성자에서 GameRecord 객체를 받아 클래스 변수로 사용할 수 있도록 한다.
- executeRace 메소드에서 반환값을 없앤다.
- executeRace 메소드에 게임 결과를 기록하는 기능을 추가한다.
- getGameResult 메소드를 추가한다.
- 우승자를 찾는 메소드를 추가한다.

Combi153 committed:
refactor: RandomNumberGenerator 클래스를 리팩토링한다

변경 내용 :
- NumberGenerator 인터페이스를 사용한다.
- RandomSingleDigitGenerator 로 이름을 수정한다.
- RandomSingleDigitGenerator 이 NumberGenerator 인터페이스의 구현체가 되도록 한다.
- RandomSingleDigitGenerator 가 0부터 9의 숫자를 무작위로 뽑는 기능을 갖도록 수정한다.

Combi153 committed:
feat: GameRecord 객체가 게임 결과를 기록하도록 구현한다

Combi153 committed:
feat: CarDto 객체를 구현한다

Combi153 committed:
refactor: GameController 가 도메인 객체들의 변경 사항을 반영하도록 리팩토링한다

Combi153 committed:
refactor: OutputView 객체가 도메인 객체의 변화를 반영할 수 있도록 리팩토링한다

Combi153 committed:
refactor: Result 객체에 우승자를 찾는 기능을 삭제한다.

Combi153 committed:
docs: 고민 및 리팩토링 내용을 추가한다

Combi153 committed:
fix: Cars 클래스의 Set 자료구조의 구현체를 변경한다

Combi153 committed:
refactor: CarDto 클래스를 데이터 저장 객체로 변경한다

변경 내용 :
CarDto 클래스의 이름을 GameResultOfCar 로 변경한다.
클래스 변수로 gameRound를 추가한다.
equals, hashCode, toString 메소드를 추가한다.

Combi153 committed:
feat: GameRecorder 클래스에 데이터를 저장하는 기능을 구현한다

Combi153 committed:
feat: GameRecordManager 클래스에 데이터 처리 기능을 구현한다

Combi153 committed:
refactor: GameSystem 클래스를 리팩토링한다

변경 내용 :
의존 객체의 변경 사항을 반영한다.
executeRace 메소드의 반복문에서 초기값을 상수로 선언해 사용한다.
반복 조건을 수정한다.
메소드의 이름을 수정한다.
우승자를 찾는 기능을 수정한다.

Combi153 committed:
refactor: OutputView 클래스를 리팩토링한다

변경 내용 :
메소드의 파라미터를 변경한다.
게임 결과를 출력하는 기능을 리팩토링한다.
- 게임 라운드별로 공백 라인을 출력하는 기능을 수정한다.
- 첫번째 게임 라운드를 찾는 기능을 추가한다.
우승자를 출력하는 기능을 리팩토링한다.
- 메소드의 이름을 수정한다.
- 메소드 간의 순서를 수정한다.

Combi153 committed:
fix: 리팩토링에 따른 변화를 적용한다

Combi153 committed:
refactor: Car 클래스의 생성자를 리팩토링한다

변경 내용 :
생성자에서 position 을 받아 생성할 수 있도록 변경한다.
CarFactory에서 DEFAULT_POSITION을 입력받아 생성된다.

변경 이유 :
테스트를 쉽게 하기 위함이다.

Combi153 committed:
fix: Car 클래스 리팩토링 결과를 반영한다

변경 내용 :
Car 객체를 생성할 때 DEFAULT_POSITION 을 전달할 수 있도록 변경한다.

Combi153 committed:
chore: Result 객체에 관한 내용을 삭제한다

Combi153 committed:
refactor: GameSystem 객체의 executeRace 메소드를 리팩토링한다

변경 내용 :
executeRace 기능에서 RandomSingleGenerator 객체를 생성하여 moveBy 메소드에 전달하던 것을 수정한다.
executeRace 를 사용하는 클라이언트 측에서 NumberGenerator 구현체를 받아 사용할 수 있도록 한다.

Combi153 committed:
test: GameSystem 클래스의 테스트 코드를 리팩토링한다

변경 내용 :
executeRace 메소드에 관한 테스트를 수정한다.
getWinnerGameResult 메소드에 관한 테스트를 추가한다.
테스트의 DisplayName 내용을 수정한다.

Combi153 committed:
chore: 메소드의 이름을 수정한다

Combi153 committed:
docs: 리팩토링 내역을 반영한다

Combi153 committed:
chore: CarsTest 의 DisplayName 내용을 수정한다

Hyeon9mak commented on docs/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## 리팩토링
+
+- [x] 매직넘버, 하드코딩 등의 상수 처리, 사용하지 않는 메소드 및 필드 삭제
+
+- [x] 메소드 및 변수 네이밍 신경쓰기
+
+- [x] 출력에 사용할 데이터를 담는 객체와 비즈니스 로직의 분리
+    - [x] Result의 비즈니스 로직 없애기
+
+- [x] Car, Cars 객체 리팩토링
+    - [x] `객체` 스러운 기능과 상태
+
+- [x] 데이터 저장 객체 생성
+
+- [x] 테스트 코드 작성
+
+- [x] 코드 및 요구사항의 변경 가능성을 고려하기
```
👍

Combi153 replied:
감사해요😎

Hyeon9mak commented on src/main/java/racingcar/domain/record/GameResultOfCar.java:
```diff
@@ -0,0 +1,54 @@
+        return Objects.hash(gameRound, carName, position);
+    }
+
+    @Override
+    public String toString() {
+        return "GameResultOfCar{" +
+                "gameRound=" + gameRound +
+                ", carName='" + carName + '\'' +
+                ", position=" + position +
+                '}';
+    }
```
getter 메서드를 통해 출력을 모두 행하는 것 같은데, `toString`을 오버라이드 한 이유가 있을까요?

Combi153 replied:
기능을 구현하던 단계에서 의도대로 코드가 동작하지 않았습니다. 그래서 `디버깅`을 위해 `toString`을 오버라이드 해 사용했어요. 다른 사람이 개발할 때도 디버깅 등 상황에서 사용할 일이 있을 것 같아 지우지 않았습니다!

Effective Java 책에서 toString의 재정의를 강조하는 글을 참조하기도 했습니다.

마땅한 이유가 될까요?

Hyeon9mak replied:
디버깅 용도였다면 훌륭한 이유입니다 😄 나중에 캐시를 위해 사용되기도 하므로, 객체가 가지고 있는 정보를 최대한 자세히 기입해놓는게 좋은 것 같아요.

Hyeon9mak commented on src/main/java/racingcar/domain/record/GameRecorder.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.Set;
+
+public class GameRecorder {
+
+    private final Set<GameResultOfCar> records;
+    private final GameRecordManager gameRecordManager;
+
+    public GameRecorder(Set<GameResultOfCar> records, GameRecordManager gameRecordManager) {
+        this.records = records;
+        this.gameRecordManager = gameRecordManager;
+    }
```
`GameRecorder`와 `GameRecordManager`의 책임은 각각 무엇인가요? 어떤 이유로 나뉘게 된 것인지 궁금하네요.

Combi153 replied:
- `GameRecorder` 는 게임 결과를 기록하는 역할을 합니다. 
- `GameRecordManager` 는 게임 결과의 기록 방식을 관리하는 역할입니다.
- 유지 보수 관점에서 차후 게임 결과 기록하는 방식이 `변화할 수 있다`고 생각 했습니다. 지금은 GameResultOfCar 클래스에 `name`, `position`, `gameRound` 를 담아서 저장하는데, 자동차에 대한 상태가 추가된다면 새로운 정보도 저장할 수 있을 것 같습니다.
- 따라서 저장하는 방식의 변화를 `GameRecorder` 객체의 외부에서 주입을 받아 사용하는 방법을 떠올렸습니다.
- 차후 리팩토링 시에 `GameRecordManager`를 인터페이스로 두고, 다른 구현체를 만들어 사용할 계획이었습니다.

혹시 이러한 방식에 대해 어떻게 생각하시나요?

Hyeon9mak replied:
설명을 듣고 나니 이해가 되는데, 그 전에는 어떤 의도인지 궁금했어요.
저는 개인적으로 '당장 분리가 필요한게 아니면 굳이 미리 분리할 필요 없다.' 고 생각하는 편인데, 
취향의 영역이라서 후추가 끌리시는 대로 해도 좋을 것 같아요 😃

Hyeon9mak replied:
대신 그러한 설계에 대해서는 PR description 에 미리 설명을 적어두면 💯  (저도 맨날 까먹어요.. ㅎㅎ;;)

Hyeon9mak commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,76 @@
+        print(Message.GAME_ROUND_INPUT_GUIDE.getMessage());
+    }
+
+    public void printResultGuide() {
+        print(Message.GAME_RESULT_GUIDE.getMessage());
+    }
```
내부에서만 사용되는 메서드 같아요.

Combi153 replied:
앗 private로 바꾸겠습니다! 메소드 순서도 아래로 내려야겠네요

Hyeon9mak reviewed (approved):
후추~ 자동차 경주 1단계 고생 많으셨습니다~

질문 주신 내용에 대해 답변과 제 생각을 적어보았어요.
2단계 미션을 진행하면서 참고해주세요 🙂

- https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103804480
- https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103804663
- https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103804984
- https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103807410
- https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103808251

그럼 2단계에서 만나요 ~ 👋
